kattelmod/updater.py

```diff
@@ -57,6 +57,8 @@ async def _run(self) -> None:
                     component._update_time = timestamp
                     component._update(timestamp)
                     component._update_time = 0.0
+                for component in self.components:
```

Comment:
Out of interest, what would happen if you merged the two loops over `self.components`?

Reply:
Splitting like this allows all the telstate requests (across all components) to be started in parallel before waiting for any of them to complete. If the loops were merged, each component's telstate transactions would need to complete before the next one's could start.

I'm not sure how much difference this makes in practice. I'd need to check whether aioredis pipelines commands.

Reply:
Being an asyncio noob, this forced me to go read up about it again :-)

I wondered about `async for` at first, but realise now that that is for async iterators, which we don't have.

It still seems to me that the for-loop will be serial instead of concurrent. I wrote a little test:
```Python
import asyncio

async def main():
    for n in range(5):
        print('init', n)
        await asyncio.sleep(n / 10)
        print('done', n)

asyncio.run(main())
```
Then I get
```
init 0
done 0
init 1
done 1
init 2
done 2
init 3
done 3
init 4
done 4
```
with longer and longer delays towards the end, as expected.

Maybe this is not representative of the kattelmod component flushes, but I would like to understand it better.

Also, what about `gather()`ing the list of components?

Reply:
The first loop doesn't actually *do* any asynchronous work: it does synchronous work (to update sensors) and schedules new asynchronous tasks (one can tell that it doesn't do any asynchronous work because there are no await statements). In fact I've realised that one issue is that all the computations in the first loop have to finish before those async tasks will get a turn to run, so computations won't overlap with telstate updates.

I don't think changing the second loop to use gather will make much difference either, because `_flush` just waits for things to complete rather than doing anything itself, and it doesn't really matter whether one waits for things in serial or parallel, you still wait until they're all finished. There are probably some subtleties around cancellation though.

I'll update things to have an async helper function that does all the work for one component, and use gather to run them all in parallel. That should address that overlap concern I had.

Reply:
See what you think now.

Reply:
Right, I get what you're saying. My example is then not quite applicable, since `sleep` means you are actively doing nothing while `_flush` passively does nothing 😆 

I like the new version!